### DISCUSSION TITLE
Time synchronization via NTP with PicoW

### DISCUSSION THREAD

**Discussion by Spart-aaa** [astlouys/Pico-Green-Clock#8] [Ideas]
Hello Andre, 
first and foremost a big thank you for the incredible amount of effort you are putting into this project and the meticulous documentation. I am a big fan of all of the useful and well thought out features you came up with and your project has a prominent spot right at my desk every day.

I am opening this thread because I wanted to ask if it would be possible for you to implement an NTP based time-sync of the RTC at some point in the future.

I bought the Pico-Green-Clock a few months ago thinking it would be an easy task to run it based on a PicoW instead of the regular Pico and have the PicoW connect to a WiFi once per hour or so to pull in an always accurate time via NTP. However due to the fact that I have no experience at all with programming microcontrollers or C I have been failing at achieving this despite the fact that the pico-examples include an NTP client but I have no idea how to get the RTC of the clock to sync with the ntp data.

Another idea based on this would be to have an additional tiny webserver running to make the clock configurable via website directly, but that would be a whole new topic on itself.

Thank you very much for any help you can provide and best regards :)

**Comment by astlouys** [comment]
Hello Spart-aaa,

(Sorry, I don't know your name). First of all, thank you for your good words. It's always interesting to know that somewhere, somehow, your work is put to contribution...

NTP update is something that I considered from the very beginning of the project. Here is an extract of the "potential improvements" that I wrote in the header of firmware source code Version 2.00: 
"- Add a Wi-Fi interface to the Pico, so that the clock can set its time by reading it from a time reference site on the Internet. (RTC IC is said to be very accurate, but I haven't tested it over a long period of time. Not sure what is the long-term precision)."

NOTE: The Pico W was not released yet, at the time of V2.00 release... That being said, I must say that I'm positively surprised by the precision of the DS3231 over a relatively long period of time. Moreover, Firmware Version 7.00 will add automatic adjustment for Daylight Saving Time / Summer Time. Those are probably two reasons why I delayed the NTP implementation, but it is still something that I want to implement (Version 7.00 also adds "UTC delta" (time difference) that will be required with NTP implementation), However, since this is a hobby, I don't want to commit to a specific timeframe (smile)... After Version 7.00, I would like to take a look on the possibility to use PWM to control the clock display brightness. This is something that I tried without success a while ago, but there were so many new things to implement at that time that I didn't spend too much time on it and I want to try again in a soon future. If I remember well, there were synchronization problems since the scan of the clock display (changing from one row to another and from one column to another) must be done when the display is "Off" and it would need to be synchronized with the PWM. But this is only preliminary observation and I may be wrong... I need to take a closer look at it...

Since I didn't search for any specific documentation so far, could you tell me where is that NTP documentation you're talking about in your note above ?

As for the web server, I must confess that I have much less interest in UI (User Interface) than micro-controllers... But since this is something that could eventually be done "independently of" (but "in coordination with") the Green Clock code, maybe another user will offer to participate to this development. It could also allow to edit the "Calendar Events" (add or remove some of them) without requiring a recompile...

As I'm very curious, here are a few questions for you:

- You mentioned "an NTP clock update once an hour or so"... Do you make a special usage of the clock that requires such a great precision ?

- May I ask in which country you live ? Do you use Daylight Saving Time / Summer Time ? If the answer is yes, would you accept to test the auto-DST feature for your country in Version 7.00 ?

- Have you installed any external device to the clock (DHT22, Infrared remote, BME280, Passive buzzer,...)

Regards,

ASL

**Comment by astlouys** [comment]
Hello Spart-aaa,

(Sorry, I don't know your name). First of all, thank you for your good words. It's always interesting to know that somewhere, somehow, your work is put to contribution...

NTP update is something that I considered from the very beginning of the project. Here is an extract of the "potential improvements" that I wrote in the header of firmware source code Version 2.00: 
"- Add a Wi-Fi interface to the Pico, so that the clock can set its time by reading it from a time reference site on the Internet. (RTC IC is said to be very accurate, but I haven't tested it over a long period of time. Not sure what is the long-term precision)."

NOTE: The Pico W was not released yet, at the time of V2.00 release... That being said, I must say that I'm positively surprised by the precision of the DS3231 over a relatively long period of time. Moreover, Firmware Version 7.00 will add automatic adjustment for Daylight Saving Time / Summer Time. Those are probably two reasons why I delayed the NTP implementation, but it is still something that I want to implement (Version 7.00 also adds "UTC delta" (time difference) that will be required with NTP implementation), However, since this is a hobby, I don't want to commit to a specific timeframe (smile)... After Version 7.00, I would like to take a look on the possibility to use PWM to control the clock display brightness. This is something that I tried without success a while ago, but there were so many new things to implement at that time that I didn't spend too much time on it and I want to try again in a soon future. If I remember well, there were synchronization problems since the scan of the clock display (changing from one row to another and from one column to another) must be done when the display is "Off" and it would need to be synchronized with the PWM. But this is only preliminary observation and I may be wrong... I need to take a closer look at it...

Since I didn't search for any specific documentation so far, could you tell me where is that NTP documentation you're talking about in your note above ?

As for the web server, I must confess that I have much less interest in UI (User Interface) than micro-controllers... But since this is something that could eventually be done "independently of" (but "in coordination with") the Green Clock code, maybe another user will offer to participate to this development. It could also allow to edit the "Calendar Events" (add or remove some of them) without requiring a recompile...

As I'm very curious, here are a few questions for you:

- You mentioned "an NTP clock update once an hour or so"... Do you make a special usage of the clock that requires such a great precision ?

- May I ask in which country you live ? Do you use Daylight Saving Time / Summer Time ? If the answer is yes, would you accept to test the auto-DST feature for your country in Version 7.00 ?

- Have you installed any external device to the clock (DHT22, Infrared remote, BME280, Passive buzzer,...)

Regards,

ASL

**Comment by Spart-aaa** [comment]
Hello Andre,
thank you very much for your fast reply, I am Lukas (If set in the user profile you can see the real name via hovering over the nickname).

Yes you are absolutely right, the DS3231 is very stable compared to a bunch of other RTCs and in my experience has less than 2 seconds of drift per half-year while sitting on my desk. And that despite the fact that the temperatures in my office are jumping daily between 17 and 22 degree Celsius depending on where I am working in a given week. The only real limit in clock precision is due to ineptitude in setting the clock straight from the start. 

Do not worry about the time frame at all, of course this is not a request but just a suggestion that only really became possible without a bigger hassle since the introduction of the PicoW. The example I was talking about can be found here [NTP_Client](https://github.com/raspberrypi/pico-examples/tree/master/pico_w/ntp_client). The same directory contains also other interesting examples for the PicoW for example a small TCP-Server or Client or a wireless access-point [PicoW Examples](https://github.com/raspberrypi/pico-examples/tree/master/pico_w) 

Yes the calendar events and in general the incredible amount of features that you implemented could really benefit from such a feature to be configured and turned off and on directly without a recompile. It would also be a neat use case for the two cores of the Pico so that one handles the actual time critical part and the other handles the configuration bits. Or a step further would even be dynamic calendar events that could be pulled/pushed from some central source for example some home automation system. I think someone is about to work on that in a new Python based version for the clock [here](https://github.com/domneedham/pico-clock-green-python) with the MQTT implementation he wants to add. (However that is of course a totally different area once you start going there.)

Regarding your questions:

- In general because of the precision of the RTC an update once at boot and then every 24 hours would probably be sufficient for most use case. The once per hour interval is inspired by how classical radio controlled clocks work with the DCF77 based synchronization. If I remember correctly from the times before everything was a computer they synchronized once per hour and sometimes even showed a little icon while the sync was in progress. A different consideration would be to not only use the Clock as an NTP-Client but also as a local NTP-Server sometime in the future so that the time-keeping in the network always remains accurate even without an active internet connection. Especially network devices and desktop PCs tend to show quite some drift in a short amount of time if not regularly synchronized with the internet. Of course normally one would deploy a dedicated NTP-Server for that normally but if the Pico-Clock already has such a good RTC why not put it to more use than just showing its own time so well.
- I live in Germany and yes we have indeed DST here (Or here it is called CEST) and I am very happy to help you test the feature from my local perspective. The time change occurs normally on the last Sunday of March at 2AM in UTC+1 or at 1AM in UTC by adding an hour and then switches back on the last Sunday of October at 3 AM UTC+2 or 1AM UTC by subtracting an hour. So I would be installing your newest version and then artificially modify the date and time to the according dates of the switch and observe what happens.
- Unfortunately I don't have any of the addons installed yet but was definitely thinking about adding the BME280 and maybe about the passive buzzer as well sometime in the future. My soldering skills are pretty bad tho so I would probably try to not modify the PCB of the clock itself but maybe instead add it to one of the GPIOs on the Pico directly if that is a possibility. 

Best regards,
Lukas

**Reply by astlouys** [reply]
Hi Lukas,

(Thank you for the hint on how to find the real name... I realized that I'll have to complete my own profile!).

More or less 5 seconds per year for a DS3231 as per your experience is not bad at all !  But I want to explore the possibilities of the new PicoW and I think NTP is a good way to start. I've taken a quick look at the NTP client (the link you provided) and it seems relatively straightforward...

I also thought about home automation, but I found the Green Clock a little limited for that purpose. I already installed a passive buzzer, DHT22, BME280 and an infrared receiver and I was thinking to add an infrared LED so that the clock could itself acts as a remote control to start / stop the TV / stereo system, change the volume up / down, etc... But there are already many cables and it become hard to maintain. I bought a few Pico-RGB-Matrix some time ago and I thought they were a better platform to start with (better display definition, one free Pico-compatible empty socket, already equipped with an IR receiver, etc...) I thought I would eventually switch to this platform for home automation (I want to begin by adding a ZigBee interface since I already have many ZigBee lights and electric sockets in the house). I may also ask a friend of mine to help me with PC board (I am not familiar with CAD tools).

You DST parameters are the same as those for Eric who lives in Spain. That's good, I'll have two testers (smile). European countries relies on UTC to decide when to change to / from Summer Time. My tests were OK, but it's always good to have more users to try... By the way, is there an English translation for "CEST" ?  I guess the ST stands for Standard Time ? You may want to follow the thread about "Daylight Saving Time / Summer Time" (in the same repository. I will post some instructions for Eric to test Version 7.00 that I should release in the very next days.

Just a quick comment about optional device... If the design of the Green Clock has not changed, those that I have use a "double-row" socket for the Pico. So, beside each pin of the Pico, there is an empty socket hole where you can relatively easily insert a cable, given that the cable size is compatible. You can take a look in the User Guide about "optional devices" to get an idea. It is still a good idea to put some solder on the cable so that it can be more easily inserted in the socket, but that's not as bad as having to solder something to the PC board.

Regards,

André

**Reply by Spart-aaa** [reply]
Hi André,

yes I was also very impressed by the performance of the DS3231, probably it's precision stems from the active temperature compensation and it's relatively isolated position on the PCB far away from other components. I would not have thought to find such a precise RTC in a package as plentiful as the clock for the price. 

Yes the idea with the universal control might be a bit hard to scale because you would need to store dozens of individual IR codes in a somewhat identifiable way so close to implementing a full on database. Also some IR-Codes can be pretty hard to record in the first place. (The one from ACs for example are notoriously large.) A ZigBee based approach would certainly be great because it opens you up to a whole new world of devices including sensors and all kinds of actuators.

The English translation for CEST is Central European Summer Time as opposed to the regular CET which is the "default" for many European countries and is the same as UTC+1. 

Yes you are absolutely right, I reopened my clock and indeed there is still a double row of sockets so attaching some regular jumper wires there is an easy task. I think I might be ordering the additional sensor and passive buzzer the next time I am going to order some electronic supplies and give it a try.

Really looking forward to giving your new release a try soon :)

Best regards,
Lukas

**Reply by astlouys** [reply]
Hi Lukas,

For the remote control, I implemented only one that I was planning to use, but theoretically, I isolated all "decoding-oriented" data in a separate file (I called it "Memorex.cpp" since the brand of my remote is Memorex). It should be possible to simply add other files (called "Samsung.cpp" of "Panasonic.cpp", etc...) to cover more remote control without having to support  them all in the final firmware (it is only required to specified the include filename for your needs at the beginning of the source code). Obviously, you would have to decode you own remote control to do so. I was planning to create another repository to put the code required for remote control decoding. I know that other repositories about that exist on GitHub, but it was also for me a learning exercise.

Let me know if you succeed in installing new devices ;-)

**Reply by astlouys** [reply]
Hi Lukas,

I hope you will upgrade to Firmware Version 9.00 now and take advantage of the NTP periodically re-sync of the Green Clock. I will close this thread now that NTP has been implemented. Let me know if you have problems with it (or any other feature).

André

**Reply by Spart-aaa** [reply]
Hello André, 
today I finally could find the time to try out the new firmware and all of the other new features. Thank you so much for putting in all the effort as always. The fact that you even added a user friendly way to enter the WiFi credentials without having to recompile really elevates the clock to something that can be gifted to friends and family or even be put into a professional setting. (Professional grade clock displays with NTP are normally a very expensive thing at that size in comparison.)

I had no issues entering my WiFi credentials and performing the synchronization accordingly. If I set a wrong time on purpose and reboot the clock everything is also resyncing perfectly fin minutes and hours wise. However I noticed a slight offset regarding the seconds. They always seem to be around 2-3 seconds off when compared to my computer, my phones or my Apple Watch as in the minute change on the Pico Clock seem to occur 2-3 seconds after all the other devices already changed. Such a small amount is certainly not a real issue but I thought I should mention it because in the code I saw that the seconds are also being actively synchronized and am wondering what might cause the difference here.

I also tried to figure out the the exact resync intervall but could not find the concrete amount in the code or the manual but maybe I am just blind. Is it the default of around once every hour that many other devices have?

Thanks again for everything and best regards,
Lukas

**Reply by astlouys** [reply]
Hello Lukas,

I'm glad you enjoy the new Firmware version !

As for the small delay, have you compared the Green Clock with a time server over the Internet ? ("UTC clock" or the like ?) Or with a GPS ? As far as I remember, when I verified, the difference was like "a few milliseconds" off (slightly Off, but by less than one second). And if I remember, except if you're part of a corporate network configuration where the syncing is important, I believe Windows resync with NTP is done once a week.

As for the resync interval, as mentioned in the User Guide, I wanted to make sure that it would be in respect with the rules usually asked from the concerned authorities. I understand your curiosity, but I don't want to elaborate too much about this (smile). Let's say that beside the small delay that you seem to have, there shouldn't be a significant drift (let me know if you have some)... The delay is slightly longer than one hour. I wouldn't like users (I'm not saying you would) to change the resync delay to a few seconds and saying they use my Firmware when getting problems as if they tried a DoS. I may care too much, but I'm allowed to ;-)

I'm happy that you had no problem entering the credentials. Another user seemed to have problem doing it. I was wondering if there was a bug in the code and / or a glitch in the User Guide.

On another page, as you indicated, I wanted to make sure that potential users without a development environment would be able to enter the credentials with patience, but without problem :-)

If you believe that it makes the Green Clock a real value for the price, you should maybe touch a word to Waveshare... I'm sure they would like selling even more clocks !  And also... it's true that it makes a nice gift...

**Comment by Spart-aaa** [comment]
Hi André, thanks for the swift reply :)

I compared it to three different computers (One Mac, one Windows and a freshly bootet Raspberry Pi), two iPhones and one Smartwatch. All of these devices run in exact synchronization. The raspberry Pi gets it’s time from a DHCP assigned local NTP server that by itself is synchronized to pool.ntp.org therefore should be the same as the source the clock is pulling from. (Unfortunately I found out that my old GPS-receiver isn’t able anymore to show the correct time since the rollover of 2019 so I can’t use that as a reference anymore.) I tried to enter the local NTP server to the clock via IP directly to make sure that not maybe that one might be providing all of my devices with an inaccurate time however the result is the same: They all switch in perfect sync and around two seconds later the clock switches as well. Maybe it is just the display being out of sync and the NTP-client itself syncing correctly?

I also tried to exclude some fluke from a stored value in the RTC and pulled the battery to make sure that it is being initialized from scratch however the result seems to always be the same. During this test I also realized that Time and Date are being pulled in perfectly fine but that the year is not being pulled from NTP which results in a wrong day being shown unless manually adjusted.

My question about the sync was mostly because I set a 5 minute offset and waited a few hours without it returning to the real time so I wondered if the sync is maybe only happening once a day. I thought that you would keep the sync at a low rate to be a good citizen and especially with the experience of the RTC being very precise there is no need to sync very often. (So no question about DOSing the trusty pool we all rely on :) )

**Reply by astlouys** [reply]
Hi Lukas,

Give me a few days to investigate further about this 2-second delay that you observed... I'll see what it looks like on my side.

Some comments about your last message:

As far as I understood, pool.ntp.org is a URL from which you get the address of one NTP server from a list of such servers. I assume that this URL (pool.ntp.org) makes some kind of "round-robin" to distribute the requests to many different NTP servers (they should all return the time with a relatively good precision), but that would also mean that you're not synchronizing with the same server all the time.

Two seconds difference seems a lot for a micro-controller running at 125 MHz (the Pico), and the Green Clock display should be updated fast enough that you shouldn't observed such a difference except in rare occasions when the clock is performing other duties in the background. As you mentioned, 2 seconds "is not a big deal" in real life, but I'm also curious to understand why...

Each time the Green Clock receives time from NTP, the real-time IC will be set with the new time (even if it is the same), including the seconds. So, that does not explain the lag. However, the year should be part of the synchronization as well. I know that the original Firmware (from Waveshare) only stored the "23" part of the year in the DS-3231. I assumed it was the way the IC worked but I admit that I didn't checked in the specs. There may be a bug in the code logic related to that, but in theory, the year is supposed to be set as well from NTP. (Obviously, if there is a bug in the code and the year is bad, there are six chances out of seven that the day-of-week will be bad as well !

As for your last paragraph... If you make a change in the clock setup, there should be a NTP resync soon after. This is because I thought that you may have changed the Timezone and / or "DST country" and if ever the case, an NTP resync will be useful since the clock will be resync according to the new DST country and Timezone. However, there is a safety in the code: if an NTP resync has been done just a few seconds earlier (that is, a few seconds before you get out of clock setup), the NTP resync will fail as the delay was too short between two NTP requests. To check if this is the reason for the lack or resync, wait 5  minutes and enter-and-exit clock setup again (you may change time to see if the resync will get it back to normal) and the NTP should make its job a few seconds later. To say it in short, exiting clock setup forces a NTP resync, but it won't work if the delay since the previous NTP request is too short (wait at least a few minutes).

ASL
